homeassistant/components/matrix.py

```diff
@@ -0,0 +1,358 @@
+        return False
+
+    def send_message_handler(service):
+        """Handle the send_message service."""
+        bot.handle_send_message(service)
+
+    hass.services.async_register(
+        DOMAIN, SERVICE_SEND_MESSAGE, send_message_handler,
+        schema=SERVICE_SCHEMA_SEND_MESSAGE)
+
+    return True
+
+
+class MatrixBot(object):
+    """The Matrix Bot."""
+
+    def __init__(self, hass, config_file, homeserver, verify_ssl,
+                 username, password, listening_rooms, commands):
+        """Set up the client."""
+        self.hass = hass
+
+        self._session_filepath = config_file
+        self._auth_tokens = self._get_auth_tokens()
+
+        self._homeserver = homeserver
+        self._verify_tls = verify_ssl
+        self._mx_id = username
+        self._password = password
+
+        self._listening_rooms = listening_rooms
+
+        # Logging in is deferred b/c it does I/O
+        self._setup_done = False
+
+        # We have to fetch the aliases for every room to make sure we don't
+        # join it twice by accident. However, fetching aliases is costly,
+        # so we only do it once per room.
+        self._aliases_fetched_for = set()
+
+        # word commands are stored dict-of-dict: First dict indexes by room ID
+        #  / alias, second dict indexes by the word
+        self._word_commands = {}
+
+        # regular expression commands are stored as a list of commands per
+        # room, i.e., a dict-of-list
+        self._expression_commands = {}
+
+        for command in commands:
+            if not command.get(CONF_ROOMS):
+                command[CONF_ROOMS] = listening_rooms
+
+            if command.get(CONF_WORD):
+                for room_id in command[CONF_ROOMS]:
+                    if room_id not in self._word_commands:
+                        self._word_commands[room_id] = {}
+                    self._word_commands[room_id][command[CONF_WORD]] = command
+            else:
+                for room_id in command[CONF_ROOMS]:
+                    if room_id not in self._expression_commands:
+                        self._expression_commands[room_id] = []
+                    self._expression_commands[room_id].append(command)
+
+        # Log in. This raises a MatrixRequestError if login is unsuccessful
+        self._client = self._login()
+
+        def handle_matrix_exception(exception):
+            """Handle exceptions raised inside the Matrix SDK."""
+            _LOGGER.error("Matrix exception:\n %s", str(exception))
+
+        self._client.start_listener_thread(
+            exception_handler=handle_matrix_exception)
+
+        def stop_client(_):
+            """Run once when Home Assistant stops."""
+            self._client.stop_listener_thread()
+
+        self.hass.bus.listen_once(EVENT_HOMEASSISTANT_STOP, stop_client)
+
+        # This potentially does a lot of I/O, so we defer it
+        hass.add_job(self._join_rooms)
```

Comment:
This won't help as Home Assistant will wait till all jobs are done before firing the start event. Instead, listne for the EVENT_HOMEASSISTANT_START and then join the rooms.

Reply:
Done